### PR TITLE
Add parents list. Add PO2, PO3, PO4. Add getting of extra sources

### PR DESCRIPTION
**Pull request by joskuijpers**

### CODE REVIEW

**Comment by joskuijpers** [comment]
Made the suggested changes. Now it is still a mess but it is less duplication

File: src/main/scala/org/codefeedr/pipeline/Job.scala [hunk 1]
```diff
@@ -40,7 +40,61 @@ abstract class Job[In <: PipelineItem : ClassTag : Manifest : FromRecord] extend
   /**
     * Use the given datastream
```
Comment (wzorgdrager) [inline src/main/scala/org/codefeedr/pipeline/Job.scala:41]: Missing doc?

File: src/main/scala/org/codefeedr/pipeline/PipelineBuilder.scala [hunk 2]
```diff
@@ -181,6 +181,7 @@ class PipelineBuilder() {
     * If the graph is not configured yet (has no nodes), the graph is switched to a DAG automatically. If it was
     * already configured as sequential, it will throw an illegal state exception.
     */
+  @deprecated
```
Comment (wzorgdrager) [inline src/main/scala/org/codefeedr/pipeline/PipelineBuilder.scala:184]: Why deprecate it if we can just delete it?

File: src/main/scala/org/codefeedr/pipeline/PipelineBuilder.scala [hunk 3]
```diff
@@ -191,17 +192,44 @@ class PipelineBuilder() {
...
+    * @tparam V Node Out
+    * @return Builder
+    */
+  def addParents[U <: PipelineItem, V <: PipelineItem](obj: PipelineObject[U, V], parents: List[PipelineObject[PipelineItem, PipelineItem]]): PipelineBuilder = {
+    if (!graph.hasNode(obj)) {
+      graph = graph.addNode(obj)
+    }
+
+    for (item <- parents) {
+      if (!graph.hasNode(item)) {
```
Comment (wzorgdrager) [inline src/main/scala/org/codefeedr/pipeline/PipelineBuilder.scala:210]: Can this be done without a for loop?

File: src/test/scala/org/codefeedr/ScratchTest.scala [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+import org.scalatest.FunSuite
+
+
+class MyTestJoiner extends Job2[StringType, StringType] {
+  override def main(source: DataStream[StringType], secondSource: DataStream[StringType]): Unit = {
+    secondSource.print()
+  }
+}
+
+class ScratchTest extends FunSuite {
```
Comment (wzorgdrager) [inline src/test/scala/org/codefeedr/ScratchTest.scala:35]: Do wo we want this class to be merged if we're not using it?

File: src/test/scala/org/codefeedr/pipeline/JobTest.scala [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  test("Job with three sources") {
+    val a = new StringSource()
+    val b = new StringSource()
+    val c = new StringSource()
+    val job = new MyJob3()
+
+    val pipeline = new PipelineBuilder()
+      .setBufferType(BufferType.Kafka)
+      .edge(a, job)
+      .extraEdge(b, job)
```
Comment (wzorgdrager) [inline src/test/scala/org/codefeedr/pipeline/JobTest.scala:95]: Isn't this method deprecated?

**Review by wzorgdrager — APPROVED** [review]
Minor feedback, looks good.

**Pull request merged**